### PR TITLE
[stable/drupal] PV hostPath

### PR DESCRIPTION
**Pull request by scottrigby**
# Initial issue

## Initial Rationale
If the ideal local development workflow is to have as much parity with Prod as possible, and if K8S (and helm/charts) is how Prod is deployed, then a local chart (deployed to minikube etc) seems to be the perfect recipe for local development. 

Charts were never intended for local development, but there's no reason to not to extend charts to fit this use case! This PR is an attempt to add this to stable/drupal as an example chart.

## Initial status
I've brainstormed this idea with @prydonius on Slack. This is as far as I was able to get using the PV route you suggested.

As I understand from our Slack chat, the next step is to somehow bind the PVC to the new PV created here. Based on PV documentation, it would seem this PV should be used since it is created to match the PVC request - but that is not happening, it's still generating a PV dynamically.

We are then supposed to use existingClaim in the chart to mount our volume. I'm also not quite clear on this step. I've gone ahead and opened this PR just to get feedback, and work out any kinks through review/discussion/changes.

I also want to note, that the diff below works perfectly fine without PVC, by setting `--set persistence.enabled=false --set hostpath=/PATH/TO/HOST/MOUNT`:
<details>

```diff
$ git diff
diff --git a/stable/drupal/templates/deployment.yaml b/stable/drupal/templates/deployment.yaml
index 4df9332..f5944ed 100644
--- a/stable/drupal/templates/deployment.yaml
+++ b/stable/drupal/templates/deployment.yaml
@@ -66,6 +66,9 @@ spec:
       {{- if .Values.persistence.enabled }}
         persistentVolumeClaim:
           claimName: {{ template "fullname" . }}-drupal
+      {{- else if .Values.hostpath }}
+        hostPath:
+          path: {{ .Values.hostpath }}
       {{- else }}
         emptyDir: {}
       {{- end }}
```

</details>
However, I am trying to get this to work with/as a PV in this PR.

Feedback and ideas very much welcome!

## Related issues
I have opened an issue in the bitnami-docker-drupal project to [Mount drupal/core just like modules and themes](https://github.com/bitnami/bitnami-docker-drupal/issues/83), so that this PR would work not only for Drupal site builders and contrib maintainers, but also for local Drupal core development using kubernetes charts.

### CODE REVIEW

**Comment by prydonius** [comment]
We definitely want to avoid creating PersistentVolume objects as part of the chart (see recommendation in https://kubernetes.io/docs/concepts/storage/persistent-volumes/#writing-portable-configuration), so I think the correct way to handle this is to:

1. Create a PV and PVC manually and externally to the chart for the hostPath mount
2. Implement the `existingClaim` value in the chart to point to the manually created PVC
3. Prevent the chart from creating a PVC when `existingClaim` is defined

An example of the `existingClaim` approach can be found here: https://github.com/kubernetes/charts/blob/master/stable/grafana/templates/deployment.yaml#L76

**Comment by prydonius** [comment]
The approach above is tedious to setup - I think it would be great to have a discussion at a future SIG-Apps meeting about what the correct way of approaching this is.

**Comment by scottrigby** [comment]
@prydonius I opened up another PR just for adding an existing pvc to the stable/drupal chart (as an example): #1084. I started a discussion over there just for that.

Update for everyone else: "Local development on Kubernetes with charts and minikube" is on the agenda for the May 22 [SIG-Apps meeting](https://github.com/kubernetes/community/tree/master/sig-apps).

**Comment by scottrigby** [comment]
Update: I added a working example of a PV and PVC file, with steps for local development, here
https://github.com/kubernetes/charts/pull/1084#issuecomment-302879867. I'll keep this PR open separately, so that we can have a place for discussing the local dev with charts+minikube idea.

Some ideas in advance of Monday. If we do want to make this less tedious, we could either:
1. Once #1084 is merged, use this PR to add a simplified version of the steps from https://github.com/kubernetes/charts/pull/1084#issuecomment-302879867 to the README (for those who want to use the chart for local development), and call it a day.
2. Or automate this option in the chart: For instance use the earlier proposed `hostPath` parameter with a note that this will automatically create a PV (mounting that `hostPath`) and a bound PVC with the `existingClaim` option. That would save users the work of properly creating the PV/PVC manually.

However it shakes out Monday, I'll either close or update this PR accordingly.

**Comment by scottrigby** [comment]
## Update

In the SIG Apps meeting Today (video recording [here](https://youtu.be/am8Jlr_yf84?list=PL69nYSiGNLP2LMq7vznITnpd2Fk1YIZF3)) there were various ideas, and a concern voiced about using the same Helm Chart for both production and lower environments (including local for development). 

The Helm/Charts maintainers made clear that stable charts are intended to be production quality/security/extensibility charts (not only examples to emulate). But also, any of the values a team may want to change between environments (size, persistence even, etc) could be changed per environment, including local env. Luckily charts are already set up for this if desired: custom values files can override the default one via `--files` (for example, one per environment), and can further be overridden by `--set` flag as needed.

## Current status

The existingClaim work is already in a standalone PR #1084 as that is already a common pattern among other stable charts. I'll keep this PR open 

## Next steps

I'll update this PR with a parameterized version of the PV/PVC I demoed Today, and only install that if a `hostPath` option is set under `persistence` in values. If so, default the Deployment YAML to use that PVC name. This will be a test to get feedback from others, and see how well it eases DX for this (or any) use case that requires an optional `hostPath`.

**Comment by scottrigby** [comment]
PR updated (builds on top of #1084 commits).

File: stable/drupal/templates/drupal-hostpath.yaml
Comment (prydonius) [inline stable/drupal/templates/drupal-hostpath.yaml]: couldn't you reuse the `drupal-pvc.yaml` and template that based on the `hostPath` value?

**Review by prydonius — CHANGES_REQUESTED** [review]
once again, sorry for taking some time to get to this. I think it will need rebasing on top of fixes from #1084

File: stable/drupal/templates/drupal-hostpath.yaml
Comment (scottrigby) [reply stable/drupal/templates/drupal-hostpath.yaml]: @prydonius Yes! I was thinking this exact thing. I will refactor to do so. Just will need to add a tiny bit of logic in the existing PVC template.

**Comment by scottrigby** [comment]
Rebased on top of fixes from #1084

**Comment by prydonius** [comment]
@scottrigby I've merged #1084, can you rebase this again on top of master?

**Comment by scottrigby** [comment]
@prydonius yep will do now. In the meantime I also updated this PR to reuse the PVC template. Rebase on the way…

**Comment by scottrigby** [comment]
@prydonius OK, rebased on top of master.
I haven't thoroughly re-tested this yet… It's a bit late here in SF, and I'm flying back to NY in the morning… 

For now, I quickly tested on Minikube (though my machine probably needs to restart, and who knows what else from Today), here are the results:
- PV/PVC are correctly created and bound
- The specified hostPath is assigned to the PV correctly, according to `kubectl describe`
- I'm seeing a warning:

> FailedSync      Error syncing pod, skipping: timeout expired waiting for volume
s to attach/mount for pod "default"/"innocent-eagle-drupal-2029729204-4l4wf". list of unattached/unmounted volumes=[drupal-data]

Not sure why this would be. However I did upgrade my minikube version to `v0.19.1` earlier Today. Maybe someone else can help test this too? 

CC @lypht @rabellamy @michelleN @ryanj @tomdavidson (the Helm hack night - and our local dev on minikube + helm breakout - was great BTW!)

**Comment by scottrigby** [comment]
I also updated the README with system compatibility and mounting steps.

**Comment by scottrigby** [comment]
@prydonius @rabellamy @michelleN @ryanj @tomdavidson @lypht: 

Regarding the system compatibility note I added to the README - this issue was finally closed in April (note the issue number!), and links to a number of recent Windows hyperv fixes: https://github.com/kubernetes/minikube/issues/2#issuecomment-301153999

And @prydonius found something I didn't notice: the linked docs refactoring also included this [Mounting Host Folders](https://github.com/kubernetes/minikube/blob/master/docs/host_folder_mount.md) page, which clarifies that the preferred way to mount into minikube is `minikube mount`. 

I'd like to talk about this all together, and see if `minikube mount` really suits our needs, since it's on a per-pod basis, which doesn't persist with K8S self-healing pods, unlike our PV/PVC on the Deployment solution, which does. Also our solution is not necessarily tied to Minikube (although it does work with it). Thoughts?

File: stable/drupal/templates/drupal-pvc.yaml
Comment (prydonius) [inline stable/drupal/templates/drupal-pvc.yaml:21]: Just a note - I found out in https://github.com/kubernetes/kubernetes.github.io/pull/3979 that this is due to k8s using the `default` storage class is one isn't defined, which is why you get a dynamic PV if you don't define this. The PV you create has no storage class, so that's why it isn't picked up. If you define a storage class `manual` for example (it can be anything) on both the PV and PVC it should bind them as you expect.

Not saying we need to change it, just wanted to explain what I learned when looking at that doc issue :)

**Review by prydonius — APPROVED** [review]
this looks good to me - do you think this is ready to merge @scottrigby?

File: stable/drupal/templates/drupal-pvc.yaml
Comment (scottrigby) [reply stable/drupal/templates/drupal-pvc.yaml:21]: @prydonius Ah, this makes sense. Good sleuthing! 🔍 :shipit:

**Comment by scottrigby** [comment]
@prydonius The functionally, yes 👍  And even with the above question, we could merge this in, and still keep future discussion open about whether `minikube mount` is robust enough to replace K8S native `hostPath` at some point (I suspect not because it's per pod, but it's at least worth discussing).

But as of my last note in https://github.com/kubernetes/charts/pull/1057#issuecomment-307946607 I think README needs another update, or review – one way or the other – about the current status of Windows support (can Windows now use `hostPath` with hyperv now or not)?

OTOH, maybe it's not our job to track that, but is just enough for us to link to the official Minikube [Mounting Host Folders](https://github.com/kubernetes/minikube/blob/master/docs/host_folder_mount.md) documentation, just so people know there are other options…

I'm trying to think through this out-loud, for transparency. How about I add that, and as long as the documentation looks good then yes it will be ready to merge.

**Comment by scottrigby** [comment]
@prydonius Tl;dr, I think I update the README now to cover all our bases. If you agree, I think this is ready.

**Comment by prydonius** [comment]
@scottrigby lgtm, I'll merge this as-is and we can always follow-up with more documentation changes later :+1: - thanks!

**Comment by prydonius** [comment]
@k8s-bot ok to test

**Pull request merged**
